sympy/combinatorics/rewritingsystem.py

```diff
@@ -139,6 +143,11 @@ def add_rule(self, w1, w2, check=False):
                 new = self.add_rule(w1, w2, check)
                 new_keys.update(new)
 
+        # Add new rules to the `new_rules` dictionary.
+        if self.reduction_automaton:
+            for new_rule in new_keys:
+                self.new_rules[new_rule] = self.rules[new_rule]
```

Comment:
This may work better directly in `_add_rule` because then you wouldn't need to iterate over `new_keys` - you could just add the rule as it's added to the `rules` dictionary.

Reply:
The tests run a tad bit faster when it's done in the `_add_rule` method. Probably because we don't use the `self.rules` dictionary there. We simply define `self.new_rules[r1] = r2`